include/silo/common/cons_list.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+//! Cons list. The tail of the list is stored as a normal reference
+//! (no reference counting). This is meant to be used with recursive
+//! algorithms to maintain a path back up.
+
+#include <functional>
+#include <memory>
+#include <optional>
+
+template <typename T>
+class ConsList {
```

Comment:
What it is used for is to keep caller context in a linked list, on the native stack; it's making use of the stack, just sort of adding a field to it.

This is an efficient idiom used with pure data structures (wide-spread in functional languages): say you have a tree data structure (e.g. think XML or HTML tree), with links from parents to children; but since it's purely functional, there are no links back from the children to the parents (this would require mutation). In an imperative world libraries typically add links to the parent into the data structure itself. But that makes the data structure non-functional, i.e. it's not possible any more to re-use a sub-tree in a different place, because the parent links would have to be modded, so it has to be copied which worsens algorithmic complexity. But if instead the *algorithms* recursing into the tree keep their own parent links, the data structure doesn't need them. And if you allocate the parent links in the native stack, performance is top.

Your implementation allocates heap memory, and copies the vector which makes it O(n^2) for the depth of the stack. Yes, this doesn't matter as long as the depth is tiny (which will be the case here if config files aren't an attack vector which they probably aren't). But I figured I'd just make the ConsList now given that my Rust code was using it, and keep the code in both languages the same. Then we can use the same ConsList when we e.g. need to recurse down lineage trees efficiently.

Reply:
In the end, this is a trade off between having less code vs some small optimization (that doesn't matter in this case - a little heap memory for storing yaml paths doesn't matter).

Reply:
I had the same thoughts but given this is now already there and can really have great additional use-cases elsewhere, where contention on memory allocation is a serious problem, I would leave this as is

Reply:
Ok, if you think we will reuse this somewhere I'm fine with leaving it in the code base.